proofs/lfsc/signatures/core_defs.plf

```diff
@@ -0,0 +1,144 @@
+; apply symbol is in turn curried when representing function applications with
+; more than 2 arguments.
+;
+; (3) We use the following style when defining function symbols X:
+;   (a) We declare a simply typed version of the symbol:
+;         `(declare f_X term)`
+;       In other words, f_X is a term, which is the function X itself.
+;   (b) We define a macro that is a fully applied application of X. For example,
+;       when X expects two arguments, then we define:
+;         `(define X (# t1 term (# t2 term (apply (apply f_X t1) t2))))`
+;       More generally, this application consists of a right-associative chain
+;       of higher-order apply and f_X.
+;       Given this definition, we may write applications of X using SMT-LIB
+;       like syntax `(X a b)`.
+;
+; (4) Like SMT-LIB, we do not distinguish formulas and terms. Instead, formulas
+; can be seen as terms of Boolean sort.
+;
+; (5) We use a type `holds` which takes a term as an argument. This is expected
+; to take a Boolean term, e.g. a formula. The meaning of `(holds t)` is that
+; `t` can be inferred in the current context.
+;
+; Note that since our signature does not strictly enforce that terms type
+; check, the meta-level argument for the soundness of the signature relies
+; on the fact that the proof rules do not introduce terms that are not well
+; sorted according to SMT-LIB. Moreover, we assume that the input formula from
+; the user contains only well sorted terms.
+
+; Sorts
+(declare sort type)
+
+; Terms
+(declare term type)
+
+; holds: term t holds, where t should have Boolean type
+(declare holds (! t term type))
+
+; function type constructor
+(declare arrow (! s1 sort (! s2 sort sort)))
+
+; higher-order apply
+(declare apply (! t1 term (! t2 term term)))
+
+; Boolean value terms
+(declare true term)
+(declare false term)
+
+; Negation
+(declare f_not term)
+(define not (# t term (apply f_not t)))
+
+; Conjunction
+(declare f_and term)
+(define and (# t1 term (# t2 term (apply (apply f_and t1) t2))))
+
+; Disjunction
+(declare f_or term)
+(define or (# t1 term (# t2 term (apply (apply f_or t1) t2))))
+
+; Implication
+(declare f_=> term)
+(define => (# t1 term (# t2 term (apply (apply f_=> t1) t2))))
+
+; Xor
+(declare f_xor term)
+(define xor (# t1 term (# t2 term (apply (apply f_xor t1) t2))))
+
+; ITE
+(declare f_ite term)
+(define ite (# c term (# t1 term (# t2 term (apply (apply (apply f_ite c) t1) t2)))))
+
+; Equality:
+(declare f_=  term)
+(define = (# t1 term (# t2 term (apply (apply f_= t1) t2))))
+
+; Disequality:
+(declare f_distinct term)
+(define distinct (# t1 term (# t2 term (apply (apply f_distinct t1) t2))))
+
+;; ---- free constants
+; A "free constant" is an input variable to an SMT-LIB problem. We identify
+; each free constant by an integer identifier (argument v:mpz) and a sort
+; (argument s:sort).
+(declare var (! v mpz (! s sort term)))
+
+;; ---- variables
+; A "bound variable" is a universally quantified variable. In this signature,
```

Comment:
you have an exists quantifier below, can't variables be existentially quantified?

Reply:
Yes, I'll fix